### PR TITLE
Fix ordering issue when artifacts exceed `batchSize`

### PR DESCRIPTION
**Pull request by gasperzgonec**
## Description

This PR fixes the issue that occurred when the amount of the items was larger than the batch size, which made the artifacts upload, but returned the artifact list ordered according to the upload time, not the repository order.
## Connected Issues

- [#ISS-212974](https://app.devrev.ai/devrev/works/ISS-212974)

## Checklist

- [x] Tests added/updated and ran with `npm run test` OR no tests needed.
- [x] Ran backwards compatibility tests with `npm run test:backwards-compatibility`.
- [ ] Tested airdrop-template linked to this PR.
- [x] Documentation updated and provided a link to PR / new docs OR `no-docs` written: `no-docs`

### CODE REVIEW

**Comment by gasperzgonec** [comment]
I'd like to get a bit of feedback.
I've created the `worker-adapter.ordering.test.ts` to be able to un-mock the `Repo` class, it would have mocked it otherwise and I couldn't figure out how to un-mock it correctly.

The code and the testing is still valid, it's just in a separate file, which might not be the best approach.

File: src/workers/worker-adapter.ordering.test.ts [hunk 1]
```diff
@@ -0,0 +1,99 @@
...
+
+import { createEvent, createItems } from '../tests/test-helpers';
+import { WorkerAdapter } from "./worker-adapter";
+import { State, createAdapterState } from "../state/state";
+import { EventType } from "../types";
+
+describe("Batch ordering", () => {
+
+  describe('should take batch ordering into account', () => {
+    it('should maintain artifact ordering when repo A has items below batch size and repo B has items above batch size', async () => {
```
Comment (radovanjorgic) [inline src/workers/worker-adapter.ordering.test.ts:36]: Kudos for following AAA, but should we rather use better names for example `ItemTypeA` or sth like `tasks`, `item-type-1`?

File: src/workers/worker-adapter.ordering.test.ts [hunk 2]
```diff
@@ -0,0 +1,99 @@
...
+        initialState: { attachments: { completed: false } },
+      });
+
+      const testAdapter = new WorkerAdapter({
+        event: mockEvent,
+        adapterState: mockAdapterState,
+        options: {
+          batchSize: 50
+        }
+      });
```
Comment (radovanjorgic) [inline src/workers/worker-adapter.ordering.test.ts:53]: Let's add more tests and consequently move these out of `it` and define at top of `describe` block if they will be common for multiple tests. Check `beforeEach` helper.

File: src/workers/worker-adapter.ordering.test.ts [hunk 3]
```diff
@@ -0,0 +1,99 @@
...
+
+      // Track the order of artifacts added to the adapter
+      const artifactOrder: string[] = [];
+      const originalArtifacts = testAdapter.artifacts;
+
+      // Override the artifacts setter to track order
+      Object.defineProperty(testAdapter, 'artifacts', {
```
Comment (radovanjorgic) [inline src/workers/worker-adapter.ordering.test.ts:60]: Can't we mock these?

File: src/workers/worker-adapter.ordering.test.ts [hunk 4]
```diff
@@ -0,0 +1,99 @@
...
+        { itemType: 'B' }
+      ]);
+
+      await testAdapter.getRepo('A')?.push(createItems(5));
+      await testAdapter.getRepo('B')?.push(createItems(105));
+
+      await testAdapter.uploadAllRepos();
+
+      const artifacts = testAdapter.artifacts;
+      console.log(artifacts);
```
Comment (radovanjorgic) [inline src/workers/worker-adapter.ordering.test.ts:83]: Remove log.

File: src/workers/worker-adapter.ordering.test.ts [hunk 5]
```diff
@@ -0,0 +1,99 @@
...
+      let firstBFound = false;
+      for (const artifact of artifacts) {
+        if (artifact.item_type === 'B') {
+          firstBFound = true;
+        } else if (artifact.item_type === 'A') {
+          // If we find an 'A' artifact after we've already found a 'B' artifact,
+          // the ordering is incorrect
+          expect(firstBFound).toBe(false);
+        }
+      }
```
Comment (radovanjorgic) [inline src/workers/worker-adapter.ordering.test.ts:96]: Can't we just check if first item is of the same type as it was in initializeRepo array. Maybe we can even check for all items (if batch size is 50, we should have 4 artifacts (A 5, B 50, B 50, B 5)?

File: src/workers/worker-adapter.ordering.test.ts [hunk 6]
```diff
@@ -0,0 +1,99 @@
...
+    });
+  });
+});
```
Comment (radovanjorgic) [inline src/workers/worker-adapter.ordering.test.ts:99]: Add more tests and cover some edge cases.. what if we push 0 items, will it show up in the final artifacts array, check how test behaves if we have single repo and so on :)

**Review by radovanjorgic — COMMENTED** [review]
Added few comments, please follow testing guidelines we defined, check how is it done for some other test files add some more tests. Also, can we move it to `worker-adapter.test.ts` or organize it somehow better if you think we need separate file for this?

Please always make sure that you did everything listed in checklist (also link with airdrop template and test).

File: src/repo/repo.ts
[the remark this answers is not in the document]
Comment (gasperzgonec) [reply src/repo/repo.ts:1]: That is true, but I'm not sure what we can do about that. These changes don't have anything to do with the memory usage of non-uploaded items.
There could be improvements, of course, and we'd need to do some research to figure out what the best solution would be.

File: src/workers/worker-adapter.ordering.test.ts [hunk 7]
```diff
@@ -0,0 +1,99 @@
...
+
+import { createEvent, createItems } from '../tests/test-helpers';
+import { WorkerAdapter } from "./worker-adapter";
+import { State, createAdapterState } from "../state/state";
+import { EventType } from "../types";
+
+describe("Batch ordering", () => {
+
+  describe('should take batch ordering into account', () => {
```
[the remark this answers is not in the document]
Comment (gasperzgonec) [reply src/workers/worker-adapter.ordering.test.ts:35]: I used to have two describe blocks inside the first one, will fix.

File: src/workers/worker-adapter.ordering.test.ts (hunk 1 above)
Comment (gasperzgonec) [reply src/workers/worker-adapter.ordering.test.ts:36]: I don't think it matters much, as these are really just placeholders, but will change for readability purposes.

File: src/workers/worker-adapter.ordering.test.ts (hunk 5 above)
Comment (gasperzgonec) [reply src/workers/worker-adapter.ordering.test.ts:96]: There already is a check if there's 4 artifact items, but I'd rather not hard-code them.
I have changed it, so that the check now checks if the ordering matches the one in the `initializeRepos` array.

File: src/workers/worker-adapter.ordering.test.ts (hunk 6 above)
Comment (gasperzgonec) [reply src/workers/worker-adapter.ordering.test.ts]: I have added a few more tests, one for empty array being pushed to the repo, one for only a single repo, and lastly, one for pushing and uploading during the process.
All pass :)

File: src/workers/worker-adapter.ordering.test.ts
[the remark this answers is not in the document]
Comment (gasperzgonec) [reply src/workers/worker-adapter.ordering.test.ts:1]: Apologies, I copied it from another test file when nothing was working, and I got it to work and never touched it to refactor. If we were overriding imports, these mocks would have to be before the imports, but in this case it works like this with no issues :)

File: src/workers/worker-adapter.ordering.test.ts (hunk 3 above)
Comment (radovanjorgic) [reply src/workers/worker-adapter.ordering.test.ts]: But what is the point of this? I see test passes even without this overridings and none of the `originalArtifacts` or `artifactOrder` aren't used in test.

**Review by radovanjorgic — APPROVED** [review]
Rebase to the latest main. Otherwise LG.

File: src/workers/worker-adapter.ordering.test.ts (hunk 3 above)
Comment (gasperzgonec) [reply src/workers/worker-adapter.ordering.test.ts]: Uh, that was an oversight. It was the old way of tracking the artifacts before making them public, but the issue with this was that the logic for adding the artifacts wasn't being tested correctly.
This is a dead code and has been removed. Thanks for pointing it out.
